pkg/istio/rbac/common/util.go

```diff
@@ -202,30 +202,75 @@ func MemberToOriginJwtSubject(member interface{}) (string, error) {
 // SPIFFE name format: <athenz domain name>/ra/<role name>
-func RoleToSpiffe(athenzDomainName string, roleName string) (string, error) {
+func RoleToSpiffe(athenzDomainName string, roleName string, enableSpiffeTrustDomain bool, systemNamespaces []string) ([]string, error) {
 	if len(athenzDomainName) == 0 {
-		return "", fmt.Errorf("athenzDomainName is empty")
+		return nil, fmt.Errorf("athenzDomainName is empty")
 	}
 	if len(roleName) == 0 {
-		return "", fmt.Errorf("roleName is empty")
+		return nil, fmt.Errorf("roleName is empty")
+	}
+
+	var namespace string
+	if containsSystemNamespace(athenzDomainName, systemNamespaces) {
+		namespace = athenz.DomainToNamespaceForSystemComponents(athenzDomainName)
+	} else {
+		namespace = athenz.DomainToNamespace(athenzDomainName)
+	}
+	principal := fmt.Sprintf("%s.%s", athenzDomainName, roleName)
+
+	spiffeNames := []string{
+		fmt.Sprintf("%s/ra/%s", athenzDomainName, roleName),
+	}
+	if enableSpiffeTrustDomain {
+		newSpiffeNames := []string{
+			fmt.Sprintf("athenz.cloud/ns/%s/ra/%s", namespace, principal),
+			fmt.Sprintf("athenz.cloud/ns/%s/ra/%s", athenzDomainName, principal),
+			fmt.Sprintf("athenz.cloud/ns/default/ra/%s", principal),
+		}
+		spiffeNames = append(spiffeNames, newSpiffeNames...)
 	}
-	spiffeName := fmt.Sprintf("%s/ra/%s", athenzDomainName, roleName)
-	return spiffeName, nil
+	return spiffeNames, nil
+}
+
+func containsSystemNamespace(domain string, systemNamespaces []string) bool {
+	for _, ns := range systemNamespaces {
+		if strings.Contains(domain, ns) {
+			return true
+		}
+	}
+	return false
 }
 
 // PrincipalToSpiffe converts the Athenz principal into a SPIFFE compliant format
 // e.g. client-domain.frontend.some-app -> client-domain.frontend/sa/some-app
-func PrincipalToSpiffe(principal string) (string, error) {
+func PrincipalToSpiffe(principal string, enableSpiffeTrustDomain bool, systemNamespaces []string) ([]string, error) {
 	if len(principal) == 0 {
-		return "", fmt.Errorf("principal is empty")
+		return nil, fmt.Errorf("principal is empty")
 	}
 	i := strings.LastIndex(principal, ".")
 	if i < 0 {
-		return "", fmt.Errorf("principal:%s is not of the format <Athenz-domain>.<Athenz-service>", principal)
+		return nil, fmt.Errorf("principal:%s is not of the format <Athenz-domain>.<Athenz-service>", principal)
 	}
+
 	memberDomain, memberService := principal[:i], principal[i+1:]
-	spiffeName := fmt.Sprintf("%s/sa/%s", memberDomain, memberService)
-	return spiffeName, nil
+	var namespace string
+	if containsSystemNamespace(memberDomain, systemNamespaces) {
+		namespace = athenz.DomainToNamespaceForSystemComponents(memberDomain)
+	} else {
+		namespace = athenz.DomainToNamespace(memberDomain)
+	}
+	spiffeNames := []string{
+		fmt.Sprintf("%s/sa/%s", memberDomain, memberService),
+	}
+	if enableSpiffeTrustDomain {
+		newSpiffeNames := []string{
+			fmt.Sprintf("athenz.cloud/ns/%s/sa/%s", namespace, principal),
```

Comment:
can you please add a special case for `user.<user-id>` -> in this case we would only need `user/sa/<user-id>` and `athenz.cloud/ns/default/sa/user.<userid>`